packages/playground-preview-worker/src/user.do.ts

```diff
@@ -158,17 +160,33 @@ export class UserSession {
 			}
 			return workerResponse;
 		}
+
 		const userSession = this.state.id.toString();
-		const worker = await request.formData();
 
-		const m = worker.get("metadata");
+		let worker: FormData;
+		try {
+			worker = await request.formData();
+		} catch (e) {
+			throw new BadUpload(`Expected valid form data`);
+		}
 
-		assert(m instanceof File);
+		const m = worker.get("metadata");
+		if (!(m instanceof File)) {
+			throw new BadUpload("Expected metadata file to be defined");
+		}
 
-		const uploadedMetadata = JSON.parse(await m.text());
+		let uploadedMetadata: UploadedMetadata;
+		try {
+			uploadedMetadata = UploadedMetadata.parse(JSON.parse(await m.text()));
```

Comment:
Out of curiosity, I seem to remember having to do this before but can't remember why this works when `UploadedMetadata.parse(await m.json())` doesn't...

Reply:
Right that's it. If you use `.json()` then you don't get access to the bad content to use in the error message.
Given the brevity of our other errors I think adding the body to the error is not necessary.